src/compose/ComposeBox.js

```diff
@@ -217,6 +220,47 @@ class ComposeBox extends PureComponent<Props, State> {
     }
   };
 
+  insertAttachment = async (attachments: DocumentPickerResponse[]) => {
+    this.setState(({ numUploading }) => ({
+      numUploading: numUploading + 1,
+    }));
+
+    const { _, auth } = this.props;
+
+    let placeholders = '';
+    for (let i = 0; i < attachments.length; i++) {
+      const fileName = attachments[i].name ?? `attachment-${i + 1}`;
+      placeholders += `${_('[Uploading {fileName}...]()', { fileName })}\n\n`;
+    }
+    this.insertMessageTextAtCursorPosition(placeholders);
+
+    for (let i = 0; i < attachments.length; i++) {
+      const fileName = attachments[i].name ?? `attachment ${i + 1}`;
+      const placeholder = _('[Uploading {fileName}...]()', { fileName });
+      try {
+        const response = await api.uploadFile(auth, attachments[i].uri, fileName);
+        const linkText = `[${fileName}](${response.uri})`;
+        if (this.state.message.indexOf(placeholder) !== -1) {
+          this.setMessageInputValue(this.state.message.replace(placeholder, linkText));
+        } else {
+          this.setMessageInputValue(`${this.state.message} ${linkText}`);
+        }
+      } catch (e) {
+        showToast(_('Uploading {fileName} unsuccessful.', { fileName }));
+        this.setMessageInputValue(
+          this.state.message.replace(
+            placeholder,
+            _('[Uploading {fileName} failed.]()', { fileName }),
+          ),
+        );
+      }
+    }
```

Comment:
Let's make this try/catch be tightly scoped around just the `await api.uploadFile` line. That way the error handling in the `catch` block gets to be focused on a very specific way things can fail, which is the API call.

The fundamental difference between an exception from the API call and an exception from any of the rest of this code is that an exception from any of the rest would represent a bug in our code. But with an API call it's part of its interface that it might throw an exception for reasons that aren't a bug in anything, just the way the network is at that moment. Generally when throwing an exception is a normal part of the interface of something, we want to handle it explicitly right there, but for exceptions that would represent a bug we want to let them propagate up to very general handlers at the outer layers of the app.

One way in particular that having this larger `try` block would be bad, if a situation ever arises where it makes a difference, is that it would mean we had a bug somewhere in the code that's trying to replace the placeholder with the link text -- and we'd be suppressing the error here, so that we didn't get a report of it in Sentry to help us find and fix it.

Reply:
I see, I was wondering why this specific code pattern --
```javascript
try {
  mayThrowError()
} catch (e) {
  // handle error
  return;
}
// handle success
```
-- was present at some places, instead of just handling success inside the try block after `mayThrowError`. Infact I changed this in https://github.com/zulip/zulip-mobile/commit/e6bcb0e8905cf02712d62f1ae40079b467b459ca thinking it was not necessary, which you also addressed in https://github.com/zulip/zulip-mobile/pull/4514#discussion_r620897833. This makes sense.

Reply:
Yep, exactly.

I've also now gone and expanded this to an entry in our style guide: d99ad3ad2